File: BOJ/14889.cpp
#include <stdio.h>
#include <vector>

using namespace std;

int n, map[25][25], result = 86796978, team[25];

int getScore(vector <int>& tt) {
	int score = 0;

	for (int i = 0; i < tt.size() - 1; i++) {
		for (int j = i + 1; j < tt.size(); j++) {
			int n1 = tt[i];
			int n2 = tt[j];
			score += map[n1][n2];
			score += map[n2][n1];
		}
	}
	return score;
}

void getResult(int index, int steam) {
	if (index >= n) {
		if (steam == n / 2) {
			vector <int> st, lt;
			for (int i = 0; i < n; i++) {
				if (team[i] == 0) st.push_back(i);
				else lt.push_back(i);
			}
			int dif = getScore(st) - getScore(lt);
			if (dif < 0) dif *= -1;
			if (dif < result) result = dif;
		}
		return;
	}
	for (int i = 0; i < 2; i++) {
		team[index] = i;
		if (i == 0) getResult(index + 1, steam + 1);
		else getResult(index + 1, steam);
	}
}

int main() {
	scanf("%d", &n);
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			scanf("%d", &map[i][j]);
		}
	}
	getResult(0, 0);
	printf("%d", result);
	return 0;
}

File: BOJ/14889/14889.py
import sys

# sys.stdin = open('BOJ/14889/example.txt')
sys.setrecursionlimit(10 ** 5)
input = sys.stdin.readline

N = int(input().rstrip())
S = [[0] * N for _ in range(N)]
scoreMap = {}
result = 99999999

def getScore(teamMembers):
  score = 0
  
  for idx, member1 in enumerate(teamMembers):
    for idx2, member2 in enumerate(teamMembers):
      if idx != idx2:
        score += S[member1][member2]

  return score

def getTeamCombination(team):
  if len(team) == N:
    if sum(team) == N / 2:
      global result

      team0 = [index for index, val in enumerate(team) if val == 0]
      team0.sort()
      team1 = [index for index, val in enumerate(team) if val == 1]
      team1.sort()

      team0Name = "".join(map(str, team0))
      team1Name = "".join(map(str, team1))

      # 이미 구한 적 있는 조합은 스킵
      if scoreMap.get(team0Name, False):
        return
      
      scoreMap[team0Name] = getScore(team0)
      scoreMap[team1Name] = getScore(team1)

      result = min(abs(scoreMap[team0Name] - scoreMap[team1Name]), result)

    return
  
  for i in range(2):
    getTeamCombination(team + [i])

for i in range(N):
  S[i] = list(map(int, input().rstrip().split()))

getTeamCombination([])
print(result)